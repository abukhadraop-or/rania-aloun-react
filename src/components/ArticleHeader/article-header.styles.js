import styled from 'styled-components';
import { ReactComponent as heart } from 'assets/heart.svg';
import { ReactComponent as emptyHeart } from 'assets/empty-heart.svg';
import colors from 'styles/colors';

export const Container = styled.div`
  display: flex;
  justify-content: space-between;
`;

export const ArticleDate = styled.p`
  color: ${colors.smokeyGrey};
  font-size: 0.8rem;
  margin: 0;
`;

export const LikeButtonContainer = styled.button`
  background-color: transparent;
  border-radius: 0.125rem;
  border: 0.063rem solid ${colors.boringGreen};
  cursor: pointer;
  display: flex;
  height: fit-content;
  padding: 0.2rem 0.5rem;
`;

export const LikeIcon = styled(heart)`
  fill: ${colors.boringGreen};
  width: 0.8rem;
`;

export const UnLikeIcon = styled(emptyHeart)`
  width: 0.8rem;
  fill: ${colors.boringGreen};
`;

export const LikesCount = styled.p`
  color: ${colors.boringGreen};
  margin: 0 0 0 0.125rem;
`;
