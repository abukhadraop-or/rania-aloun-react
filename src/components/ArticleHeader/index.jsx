import React, { useState } from 'react';
import PropTypes from 'prop-types';
import UserData from 'components/UserData';
import updateLikesService from 'services/update-likes';
import {
  LikeIcon,
  Container,
  UnLikeIcon,
  LikesCount,
  LikeButtonContainer,
} from './article-header.styles';

/**
 * Component holding the article card header.
 *
 * @param {Object} props Props passed to ArticleHeader Container.
 * @param {Object[]} article Passed article array holding it's publisher data with the likes of the article.
 * @param {func} onLike Sends the clicked article's id to the likes handler.
 *
 * @return {JSX.Element} Container holding the article header, publisher data and likes.
 */
function ArticleHeader({ id, liked, userName, publishDate }) {
  const [isLiked, setLike] = useState(false);
  const [count, setCount] = useState(liked);

  const handleLike = async () => {
    const trigger = isLiked ? -1 : 1;
    await updateLikesService(id, count + trigger);
    setCount(count + trigger);
    setLike(!isLiked);
  };

  return (
    <Container>
      <UserData userName={userName} publishDate={publishDate} />
      <LikeButtonContainer onClick={handleLike}>
        {isLiked ? <LikeIcon /> : <UnLikeIcon />}
        <LikesCount>{count}</LikesCount>
      </LikeButtonContainer>
    </Container>
  );
}

ArticleHeader.propTypes = {
  id: PropTypes.number.isRequired,
  liked: PropTypes.number.isRequired,
  userName: PropTypes.string.isRequired,
  publishDate: PropTypes.string.isRequired,
};
export default ArticleHeader;
