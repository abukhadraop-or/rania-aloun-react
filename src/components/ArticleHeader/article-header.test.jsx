import React from 'react';
import Enzyme, { shallow } from 'enzyme';
import Adapter from '@wojtekmaj/enzyme-adapter-react-17';
import ArticleHeader from '.';
import {
  LikeButtonContainer,
  LikeIcon,
  LikesCount,
  UnLikeIcon,
} from './article-header.styles';

Enzyme.configure({ adapter: new Adapter() });

describe('<ArticleHeader />', () => {
  const wrapper = shallow(
    <ArticleHeader
      article={{
        liked: 5,
      }}
    />
  );

  it('renders correctly', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('renders liked icon when liked button is clicked', () => {
    const likeButton = wrapper.find(LikeButtonContainer);
    likeButton.simulate('click');
    const likeIcon = wrapper.find(LikeIcon);
    const unLikeIcon = wrapper.find(UnLikeIcon);
    const likesCount = wrapper.find(LikesCount);

    expect(likeIcon.exists()).toBeTruthy();
    expect(unLikeIcon.exists()).toBeFalsy();
    expect(likesCount.text()).toEqual('6');
  });

  it('renders unliked icon when liked button is toggled', () => {
    const LikeButton = wrapper.find(LikeButtonContainer);
    LikeButton.simulate('click');
    const likeIcon = wrapper.find(LikeIcon);
    const unLikeIcon = wrapper.find(UnLikeIcon);
    const likesCount = wrapper.find(LikesCount);

    expect(likeIcon.exists()).toBeFalsy();
    expect(unLikeIcon.exists()).toBeTruthy();
    expect(likesCount.text()).toEqual('5');
  });
});
