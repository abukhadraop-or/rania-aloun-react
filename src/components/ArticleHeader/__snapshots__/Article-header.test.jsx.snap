// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ArticleHeader /> renders correctly 1`] = `ShallowWrapper {}`;
